### ISSUE TITLE
Customizable Default Configuration for Storage Class in User Clusters

### ISSUE THREAD

**Issue by csengerszabo** [kubermatic/kubermatic#13258]
### Description of the feature you would like to add / User story

As a KKP admin
I want the ability to update default configurations within user clusters, particularly for storage classes
In order to customize their behavior according to our requirements.

### Solution details

- As a user, I should be able to update the configuration of the storage class in a user cluster.
  - Particularly:
    - `allowVolumeExpansion` should be configurable and set to `true`.
    - `reclaimPolicy` should be configurable and set to `Retain`.
- The solution should allow for easy extension to accommodate potential future needs for configuring other default settings for user clusters
- The configuration of defaults is applied consistently to every newly created user cluster

### Alternative approaches

- reconfigure the defaults, at least for the storageclass for every user cluster being created

### Use cases

- Updating the default behaviour of the default storage class of the user clusters

### Additional information

- By default, if not explicitly configured, `allowVolumeExpansion` is deactivated and reclaimPolicy is set to `delete`

**Comment by embik** [comment]
This issue needs clarification: Is it _end users_ (people creating clusters) or _platform admins_ (changing the defaults of what is deployed) that should be able to change this? The first section says _As a KKP admin_, but the solution details say _As a user_. End users should just deploy their own `StorageClasses` if the defaults don't suit them.

**Comment by judge-red** [comment]
If I may chime in as a platform admin: we simply buy a custom addons image with our own storage classes included and we then overwrite the addons that are installed through the KubermaticConfiguration, i.e. we removed the `default-storage-class` addon from the list and added our own. So this is already possible.

I obviously wouldn't be opposed to having an easier solution that doesn't involve building our own addons image and overwriting the addons list. But imho only being able to configure the parameters Csenger mentioned wouldn't be helpful. We'd need to overwrite the default storage classes, probably through the KubermaticConfiguration, completely. For example we also set `.parameters.type` and we have 4 instead of 2 StorageClasses, because we provide them with both `reclaimPolicy: Retain` and `reclaimPolicy: Delete` as different end users have different needs.

**Comment by steled** [comment]
@embik my 5 cents: 
For _End Users_ this is clear, they just deploy their own `StorageClass`.
For us it is more important from _KKP Admin_ point of view.

I'm with @judge-red. It would be great to have an easier solution than building a custom addon.

**Comment by xrstf** [comment]
Is this something that could be solved by using the Applications feature in KKP?

**Comment by steled** [comment]
Hi @xrstf,

IIRC, the application feature needs to be enabled separately for each user cluster.
For us, it would be great if we could generally override or set new defaults.

**Comment by judge-red** [comment]
@xrstf I think if it was possible to mandate certain applications, like we can with addons, that could work. But I don't think that's currently possible, is it?

That said, we also don't wan't the default StorageClasses defined in the KKP addon, as they're non-functional for us. So we'd still need to overwrite the `defaultManifests` of the addons to exclude this, which is a pain point, unless these are deployed differently, or we can control deployment in a different way.

**Comment by xrstf** [comment]
You're right, but overriding the defaultManifests should at least be a bit simpler than maintaining your own set of addons. Apps would make that maintenance a bit easier was my thinking.

**Comment by judge-red** [comment]
Simpler yes. But just recently we ran into the issue that addons were removed from KKP and we missed that we should remove them from the `defaultManifests`. It was a quick fix, but it's still something we need to maintain and for which it's hard to monitor upstream changes (i.e. changes in the default `defaultManifests`).

**Comment by csengerszabo** [comment]
Would be better to configure StorageClass in the datacenter level.

**Comment by steled** [comment]
We are facing another situation where this is needed.
Currently all of the VMDKs created via the default storageclass from vSphere CSI in user cluster:

```sh
$ k get sc
NAME                    PROVISIONER              RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
vsphere-csi (default)   csi.vsphere.vmware.com   Delete          Immediate           false                  300d

$ k get sc vsphere-csi -oyaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  annotations:
    storageclass.beta.kubernetes.io/is-default-class: "true"
  labels:
    kubermatic-addon: default-storage-class
    kubernetes.io/cluster-service: "true"
  name: vsphere-csi
allowVolumeExpansion: true
provisioner: csi.vsphere.vmware.com
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

are provisioned in thin mode.
Since VMware vSphere Container Storage Plug-in 3 and vSphere 8.0 Update 1 it is possible to provision VMDKs in thick mode (see: [Support for thick provisioning and format conversion from thin and thick on VMFS datastores](https://techdocs.broadcom.com/us/en/vmware-cis/vsphere/container-storage-plugin/3-0/release-notes/vmware-vsphere-container-storage-plugin-30-release-notes.html#:~:text=Support%20for%20thick%20provisioning%20and%20format%20conversion%20from%20thin%20and%20thick%20on%20VMFS%20datastores.))

Unfortunately this can only be achieved via [storage policies with thick volume requirements](https://techdocs.broadcom.com/us/en/vmware-cis/vsphere/container-storage-plugin/3-0/release-notes/vmware-vsphere-container-storage-plugin-30-release-notes.html#:~:text=storage%20policies%20with%20thick%20volume%20requirements).

So my plan would be to create such a storage policy in vCenter and update the previously mentioned storageclass with this storageplicy (which would something look like below).

```sh
$ k get sc
NAME                          PROVISIONER              RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
vsphere-csi-thick (default)   csi.vsphere.vmware.com   Retain          Immediate           true                   7m58s

$ k get sc vsphere-csi-thick
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  annotations:
    storageclass.kubernetes.io/is-default-class: "false"
  labels:
    kubermatic-addon: default-storage-class
    kubernetes.io/cluster-service: "true"
  name: vsphere-csi-thick
allowVolumeExpansion: true
parameters:
  storagepolicyname: kubernetes-thick
provisioner: csi.vsphere.vmware.com
reclaimPolicy: Retain
volumeBindingMode: Immediate
```
